client/src/console.rs

```diff
@@ -192,17 +361,103 @@ impl InputWindow {
         self.screen.resize(w, 1)
     }
 
-    pub fn set_content(&mut self, content: &str) {
-        self.dirty = self.input_buffer != content;
-        self.input_buffer = content.to_string();
+    // pub fn set_content(&mut self, content: &str) {
+    //     self.dirty = self.input_buffer != content;
+    //     self.input_buffer = content.to_string();
+    //     self.cursor_pos = self.cursor_pos.clamp(0, self.input_buffer.len());
+    // }
+
+    pub fn get_input_buffer(&self) -> &String {
+        &self.input_buffer
     }
 
-    pub fn draw(&mut self) -> &Screen {
+    pub fn clear_input_buffer(&mut self) {
+        self.dirty = true;
+        self.input_buffer = String::new();
+        self.cursor_pos = 0;
+    }
+
+    pub fn put_char(&mut self, chr: char) {
+        self.dirty = true;
+        let mut new_buffer = self
+            .input_buffer
+            .chars()
+            .take(self.cursor_pos)
+            .collect::<String>();
+        new_buffer.push(chr);
+        new_buffer.push_str(
+            &self
+                .input_buffer
+                .chars()
+                .skip(self.cursor_pos)
+                .collect::<String>(),
+        );
+        self.input_buffer = new_buffer;
+        self.move_cursor(1);
+    }
+
+    pub fn remove_char(&mut self, amount: i32) {
+        match amount.cmp(&0) {
+            Ordering::Greater => {
+                self.dirty = true;
+                let mut new_buffer = self
+                    .input_buffer
+                    .chars()
+                    .take((self.cursor_pos as i32 - amount) as usize)
+                    .collect::<String>();
+                new_buffer.push_str(
+                    &self
+                        .input_buffer
+                        .chars()
+                        .skip(self.cursor_pos)
+                        .collect::<String>(),
+                );
+                self.input_buffer = new_buffer;
+                self.move_cursor(-amount);
+            }
+            Ordering::Less => {
+                self.dirty = true;
+                let mut new_buffer = self
+                    .input_buffer
+                    .chars()
+                    .take(self.cursor_pos)
+                    .collect::<String>();
+                new_buffer.push_str(
+                    &self
+                        .input_buffer
+                        .chars()
+                        .skip((self.cursor_pos as i32 - amount) as usize)
+                        .collect::<String>(),
+                );
+                self.input_buffer = new_buffer;
+            }
+            Ordering::Equal => {}
+        }
+    }
```

Comment:
I would suggest replacing this with something like:
```rust
if amount == 0 {
    return;
}
self.dirty = true;
let off_l = amount.max(0) as usize; // offset to the left from cursor, `positive` or 0
let off_r = amount.min(0).abs() as usize; // offset to the right from cursor,  `-negative` or 0
let pos_l = self.cursor_pos.saturating_sub(off_l);
self.input_buffer = String::from(&self.input_buffer[0..pos_l]) + &self.input_buffer[pos_r..]; // this skips the cursor +/- offsets
self.move_cursor(-amount.max(0));
```
It's a bit cryptic, but more concise, less repeating code and more performant (I think).
Thoughts?
(I'm 99% sure it works as it should)